pkg/controller/statusmanager/pod_status.go

```diff
@@ -61,17 +63,135 @@ type deploymentState struct {
 	LastChangeTime time.Time
 }
 
-// SetFromPods sets the operator Degraded/Progressing/Available status, based on
+func FindStatusCondition(conditions []corev1.NodeCondition, conditionType corev1.NodeConditionType) *corev1.NodeCondition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+func SetNodeCondition(conditions *[]corev1.NodeCondition, newCondition corev1.NodeCondition) {
+	if conditions == nil {
+		conditions = &[]corev1.NodeCondition{}
+	}
+	existingCondition := FindStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		*conditions = append(*conditions, newCondition)
+		return
+	}
+
+	if existingCondition.Status != newCondition.Status {
+		existingCondition.Status = newCondition.Status
+		existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+	}
+
+	existingCondition.Reason = newCondition.Reason
+	existingCondition.Message = newCondition.Message
+}
+
+func (status *StatusManager) setNodeNetworkUnavailable(nodeName string, ready bool, reason string, message string) {
+	if ready == true {
+		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to true for node %s", nodeName))
+	} else {
+		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to false for node %s", nodeName))
+	}
+	node := &corev1.Node{}
+	err := status.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Failed to get node %s", nodeName))
+		return
+	}
+	networkUnavailableCondition := corev1.ConditionFalse
+	if ready == false {
+		networkUnavailableCondition = corev1.ConditionTrue
+		reason += "NotReady"
+	} else {
+		reason += "Ready"
+		message = "NSX node agent is running"
+	}
+	condition := corev1.NodeCondition{
+		Type:               "NetworkUnavailable",
+		Status:             networkUnavailableCondition,
+		LastHeartbeatTime:  metav1.Now(),
+		LastTransitionTime: metav1.Now(),
+		Reason:             reason,
+		Message:            message,
+	}
+	SetNodeCondition(&node.Status.Conditions, condition)
+	// TODO: confirm if we need to add logic to avoid the duplicate update the same condition
+	err = status.client.Status().Update(context.TODO(), node)
+	if err != nil {
+		log.Error(err, "Failed to update node condition")
+	} else {
+		log.Info("Updated node condition")
+	}
+}
+
+// Get the pod status from API server
+func (status *StatusManager) SetFromPodsForNodeCondition() {
+	status.Lock()
+	defer status.Unlock()
+
+	pods := &corev1.PodList{}
+	err := status.client.List(context.TODO(), pods, client.MatchingLabels{"component": operatortypes.NsxNodeAgentContainerName})
```

Comment:
Is it possible to avoid extra K8s API call, I mean can we leverage existing pod status analysis from `SetFromPodsForOverall`?

Reply:
I'm afraid that we cannot leverage SetFromPodsForOverall because it only got the deployment/ds state, while the deatiled pod state we want is in the pod list and we cannot read them from deployment/ds. e.g. openshift clusteroperator only should know how many node-agents are ready but for NetworkUnavailbe condition we have to know the ready/unready pods' locations. So we have to additionally get the pod list.